src/generic/listctrl.cpp

```diff
@@ -2739,50 +2832,123 @@ bool wxListMainWindow::ScrollList(int WXUNUSED(dx), int dy)
 // keyboard handling
 // ----------------------------------------------------------------------------
 
+// Helper function which handles items selection correctly and efficiently. and
+// simply, mimic the wxMSW behaviour. And The benefit of this function is that it
+// ensures that wxEVT_LIST_ITEM_{DE}SELECTED events are only generated for items
+// freshly (de)selected (i.e. items that have really changed state). Useful for
+// multi-selection mode when selecting using mouse or arrows with Shift key down.
+
+void wxListMainWindow::ChangeHighlightedLines(size_t oldCurrent, size_t newCurrent)
+{
+    // Refresh the old/new focus to remove/add it
+    RefreshLine(oldCurrent);
+    RefreshLine(newCurrent);
+
+    // [Case 1] oldCurrent between pivot and newCurrent:
```

Comment:
Thanks a lot for writing these comments!

However I think they would be even more helpful if we formulated them in terms of position of `newCurrent` relatively to the other two, because `old` and `anchor` (== pivot) is what we have right now, while `new` is what we should have, so for me it's simpler to decompose the situations in the following cases, assuming, without loss of generality, that `old < anchor`:

1. `new < old < anchor`
1. `old < new < anchor`
1. `old < anchor < new`

This also makes it totally clear that the 3 cases are exclusive, as `new` must fall either before `old`, between `old` and `anchor`, or after `anchor`.

I _think_ that your logic is also correct, but it seems to be much more difficult to convince myself of it, so IMHO reformulating it like above would be better.